### PR TITLE
Remove software model overhead

### PR DESCRIPTION
**Pull request by oehme**
This removes the overhead of software model bridging for most builds. We now lazily initialize the bridging when a user explicitly uses the `model {}` block or applies a rules plugin. This saves anywhere from 5-10% in our configuration performance scenarios.

### CODE REVIEW

**Comment by oehme** [comment]
There are still a bunch of failures because a lot of tests use `addPlaceholderAction` for some reason. But I think this is ready for some feedback regardless.

**Comment by big-guy** [comment]
@oehme We used to do the same thing for clean/assemble/build tasks, but we deprecated that in 2.4 and removed it in 3.0.  I think we should just do the same for the remaining `addPlaceholder` tasks because they all fit into the mold of "these are things that Gradle provides OotB".  

We used to not automatically create a wrapper task, so you had to add your own. I guess we could also special case this small set of task names if we wanted to, deprecate the behavior and break it in 5.0.

**Comment by oehme** [comment]
Yeah, that's what I'm doing now - You'll get a deprecation warning if your task clashes with a placeholder. Thanks for the pointer to the previous deprecation.

**Comment by oehme** [comment]
@marcphilipp Tests are finally passing, this is ready for your review.

**Comment by oehme** [comment]
The synthetic test for [creating many tasks](https://builds.gradle.org/viewLog.html?buildId=12511624&tab=buildResultsDiv&buildTypeId=Gradle_Util_Performance_AdHocPerformanceScenarioLinux) is about 11% faster.
The [Gradle build](https://builds.gradle.org/viewLog.html?buildId=12511623&tab=buildResultsDiv&buildTypeId=Gradle_Util_Performance_AdHocPerformanceScenarioLinux) and [largeAndroidProject](https://builds.gradle.org/viewLog.html?buildId=12511622&tab=buildResultsDiv&buildTypeId=Gradle_Util_Performance_AdHocPerformanceScenarioLinux) are both about 7% faster.

**Comment by ldaley** [comment]
> Oh dear, so these placeHolderActions are actually different from createLater tasks, because they can be overwritten by the user.

IIRC, the placeholder mechanism works this way because for a long time there was no built-in `wrapper` task, and when we wanted to add one we had to deal with the backwards compatibility issue of a _lot_ of builds already declaring their own.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskContainer.java [hunk 1]
```diff
@@ -182,8 +179,11 @@ private static void setIfNull(Map<String, Object> map, String key, Object defaul
     private <T extends Task> T addTask(T task, boolean replaceExisting) {
         String name = task.getName();
 
-        if (placeholders.remove(name)) {
-            modelNode.removeLink(name);
+        if (placeholders.remove(name) != null) {
+            if (modelNode != null) {
+                modelNode.removeLink(name);
+            }
+            warnAboutPlaceholderDeprecation(name);
```
[the remark this answers is not in the document]
Comment (oehme) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskContainer.java]: Done

File: subprojects/diagnostics/src/integTest/groovy/org/gradle/api/reporting/model/ModelReportIntegrationTest.groovy [hunk 2]
```diff
@@ -27,6 +27,21 @@ class ModelReportIntegrationTest extends AbstractIntegrationSpec {
         when:
         run "model"
 
+        then:
+        def modelReportOutput = ModelReportOutput.from(output)
+        modelReportOutput.hasNodeStructure({
+            model() {
```
[the remark this answers is not in the document]
Comment (oehme) [reply subprojects/diagnostics/src/integTest/groovy/org/gradle/api/reporting/model/ModelReportIntegrationTest.groovy]: I went for the bridge on report option.

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/PluginsComponentReportIntegrationTest.groovy [hunk 3]
```diff
@@ -19,7 +19,7 @@ import org.gradle.api.reporting.components.AbstractComponentReportIntegrationTes
 
 class PluginsComponentReportIntegrationTest extends AbstractComponentReportIntegrationTest {
 
-    def "shows details of legacy Java project"() {
+    def "does not show legacy Java projects if software model isn't used"() {
```
[the remark this answers is not in the document]
Comment (oehme) [reply subprojects/plugins/src/integTest/groovy/org/gradle/java/PluginsComponentReportIntegrationTest.groovy]: I went for the bridge on report option.

**Pull request merged**